### PR TITLE
wait on server to become ready

### PR DESCRIPTION
**Pull request by jstroem**
Fixes: #184

### CODE REVIEW

File: vscode/src/lspClient.ts [hunk 1]
```diff
@@ -113,23 +113,17 @@ function startToitLsp(_: ExtensionContext,
...
-    isReady = true;
-  }, () => {
-    startFail = true;
+  return new Promise((resolve,reject) => {
+    setTimeout(() => {
+      // If the start failed, then VSCode already reported an error message.
+      // We could add an additional one to point the user to the trouble-shooting guide.
+      reject("The Language Server is not responding. Consult the documentation.");
+    }, 3000);
+    // reject("The Language Server is not responding. Consult the documentation.");
```
Comment (floitsch) [inline vscode/src/lspClient.ts:122]:
```suggestion
```

File: vscode/src/lspClient.ts [hunk 2]
```diff
@@ -113,23 +113,17 @@ function startToitLsp(_: ExtensionContext,
...
-  let startFail = false;
-  setTimeout(() => {
-    // If the start failed, then VSCode already reported an error message.
-    // We could add an additional one to point the user to the trouble-shooting guide.
-    if (!startFail && !isReady) {
-      Window.showErrorMessage("The Language Server is not responding. Consult the documentation.");
-    }
-  }, 3000);
-  result.onReady().then(() => {
-    // At this point the language server responded to requests (`initialize`), and
```
Comment (floitsch) [inline vscode/src/lspClient.ts:126]: Why remove the comment?

File: vscode/src/lspClient.ts [hunk 3]
```diff
@@ -113,23 +113,17 @@ function startToitLsp(_: ExtensionContext,
...
-    //   is thus running.
-    isReady = true;
-  }, () => {
-    startFail = true;
+  return new Promise((resolve,reject) => {
```
Comment (floitsch) [inline vscode/src/lspClient.ts:116]:
```suggestion
  return new Promise((resolve, reject) => {
```

File: vscode/src/lspClient.ts [hunk 4]
```diff
@@ -113,23 +113,17 @@ function startToitLsp(_: ExtensionContext,
...
+    setTimeout(() => {
+      // If the start failed, then VSCode already reported an error message.
+      // We could add an additional one to point the user to the trouble-shooting guide.
+      reject("The Language Server is not responding. Consult the documentation.");
```
Comment (floitsch) [inline vscode/src/lspClient.ts:120]: So you are relying on the fact that a Promise can't be completed more than once?
I would add a comment. I couldn't find good documentation for it (whether it's allowed or good practice), but it seems t work.

**Review by floitsch — APPROVED** [review]
LGTM.

File: vscode/src/lspClient.ts [hunk 5]
```diff
@@ -204,6 +198,14 @@ export function activateLsp(context: ExtensionContext, lspCommand: Array<string>
     }
   }
 
+  async function didOpenTextDocument(document: TextDocument): Promise<void> {
+    try {
+      await didOpenTextDocumentThrow(document);
+    } catch(e) {
+      Window.showErrorMessage("" + e);
```
Comment (floitsch) [inline vscode/src/lspClient.ts:205]: Other question: what would happen if we informed vscode about the error?
Would it be ignored? ...

File: vscode/src/lspClient.ts (hunk 4 above)
Comment (jstroem) [reply vscode/src/lspClient.ts:120]: It is fairly common way to make a timeout on a promise and we use it many places.

File: vscode/src/lspClient.ts (hunk 5 above)
Comment (jstroem) [reply vscode/src/lspClient.ts:205]: The error is ignored, I don't think vscode handles errors on events.

**Pull request merged**